### PR TITLE
GH-3570 LMDB store multi-threading

### PR DESCRIPTION
**Pull request by kenwenzel**
GitHub issue resolved: #3570 

Briefly describe the changes proposed in this PR:

Rewrite LmdbSailStore and introduce the possibility to insert triples on a second thread.

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md) for more details):

 - [x] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [x] I've added tests for the changes I made
 - [x] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [x] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits where necessary 
 - [x] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change

### CODE REVIEW

File: core/sail/lmdb/src/test/java/org/eclipse/rdf4j/sail/lmdb/benchmark/QueryBenchmarkComplex.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
+/*******************************************************************************
+ * Copyright (c) 2021 Eclipse RDF4J contributors.
```
Comment (jeenbroekstra) [inline core/sail/lmdb/src/test/java/org/eclipse/rdf4j/sail/lmdb/benchmark/QueryBenchmarkComplex.java:2]: Happy new year :)

File: core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/LmdbSailStore.java [hunk 2]
```diff
@@ -56,6 +56,89 @@
...
 
+	private final ExecutorService tripleStoreExecutor = Executors.newCachedThreadPool();
+	private final CircularBuffer<Operation> opQueue = new CircularBuffer<>(1024);
+	private volatile Throwable tripleStoreException;
+	private final AtomicBoolean running = new AtomicBoolean(false);
+	private boolean multiThreadingActive;
+	private volatile boolean asyncTransactionFinished;
+	private volatile boolean nextTransactionAsync;
+
+	private boolean enableMultiThreading = true;
```
Comment (jeenbroekstra) [inline core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/LmdbSailStore.java:67]: Is this something you want to make configurable?

**Review by jeenbroekstra — COMMENTED** [review]
I can't pretend to fully follow it but from what I do understand it looks really good.

**Comment by kenwenzel** [comment]
The following benchmarks were run on a VM with networked storage (HDD RAID).
The performance of larger transactions increases by up to 50%.
Only the smallest transactions with one statement per transaction loose some performance by around 5 - 10%.

```
# w/o multi-threading
Benchmark                                                       Mode  Cnt      Score     Error  Units
TransactionsPerSecondBenchmark.largerTransaction               thrpt    5     24.377 ±   1.117  ops/s
TransactionsPerSecondBenchmark.largerTransactionLevelNone      thrpt    5     27.533 ±   0.135  ops/s
TransactionsPerSecondBenchmark.mediumTransactionsLevelNone     thrpt    5  11059.920 ± 822.723  ops/s
TransactionsPerSecondBenchmark.transactions                    thrpt    5   9973.476 ± 207.579  ops/s
TransactionsPerSecondBenchmark.transactionsLevelNone           thrpt    5  17916.608 ± 297.140  ops/s
TransactionsPerSecondBenchmark.veryLargerTransactionLevelNone  thrpt    5      0.219 ±   0.001  ops/s
TransactionsPerSecondBenchmarkFoaf.transaction100kx            thrpt    5      1.561 ±   0.087  ops/s
TransactionsPerSecondBenchmarkFoaf.transaction100kxLevelNone   thrpt    5      1.914 ±   0.132  ops/s
TransactionsPerSecondBenchmarkFoaf.transaction10kx             thrpt    5     16.085 ±   0.945  ops/s
TransactionsPerSecondBenchmarkFoaf.transaction10kxLevelNone    thrpt    5     17.421 ±   1.077  ops/s
TransactionsPerSecondBenchmarkFoaf.transaction10x              thrpt    5   4512.442 ± 112.638  ops/s
TransactionsPerSecondBenchmarkFoaf.transaction10xLevelNone     thrpt    5   6014.316 ± 179.311  ops/s
TransactionsPerSecondBenchmarkFoaf.transaction1x               thrpt    5   8870.942 ±  55.528  ops/s
TransactionsPerSecondBenchmarkFoaf.transaction1xLevelNone      thrpt    5  14900.799 ± 429.247  ops/s

# w multi-threading
Benchmark                                                       Mode  Cnt      Score     Error  Units
TransactionsPerSecondBenchmark.largerTransaction               thrpt    5     33.790 ±   5.030  ops/s
TransactionsPerSecondBenchmark.largerTransactionLevelNone      thrpt    5     49.999 ±   4.165  ops/s
TransactionsPerSecondBenchmark.mediumTransactionsLevelNone     thrpt    5  11947.161 ± 412.982  ops/s
TransactionsPerSecondBenchmark.transactions                    thrpt    5   9421.710 ± 260.423  ops/s
TransactionsPerSecondBenchmark.transactionsLevelNone           thrpt    5  16480.312 ± 920.694  ops/s
TransactionsPerSecondBenchmark.veryLargerTransactionLevelNone  thrpt    5      0.306 ±   0.048  ops/s
TransactionsPerSecondBenchmarkFoaf.transaction100kx            thrpt    5      2.022 ±   0.053  ops/s
TransactionsPerSecondBenchmarkFoaf.transaction100kxLevelNone   thrpt    5      3.317 ±   0.288  ops/s
TransactionsPerSecondBenchmarkFoaf.transaction10kx             thrpt    5     22.731 ±   0.571  ops/s
TransactionsPerSecondBenchmarkFoaf.transaction10kxLevelNone    thrpt    5     27.129 ±   1.473  ops/s
TransactionsPerSecondBenchmarkFoaf.transaction10x              thrpt    5   4796.498 ± 148.201  ops/s
TransactionsPerSecondBenchmarkFoaf.transaction10xLevelNone     thrpt    5   6479.174 ± 244.761  ops/s
TransactionsPerSecondBenchmarkFoaf.transaction1x               thrpt    5   8513.821 ± 277.690  ops/s
TransactionsPerSecondBenchmarkFoaf.transaction1xLevelNone      thrpt    5  13674.143 ± 775.809  ops/s
```

File: core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/LmdbSailStore.java (hunk 2 above)
Comment (kenwenzel) [reply core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/LmdbSailStore.java:67]: I also thought about making this configurable. When looking at the benchmarks below I am unsure if this is really necessary.

File: core/sail/lmdb/src/test/java/org/eclipse/rdf4j/sail/lmdb/benchmark/QueryBenchmarkComplex.java (hunk 1 above)
Comment (kenwenzel) [reply core/sail/lmdb/src/test/java/org/eclipse/rdf4j/sail/lmdb/benchmark/QueryBenchmarkComplex.java]: Happy new year 🧨 🎆

File: core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/LmdbSailStore.java (hunk 2 above)
Comment (jeenbroekstra) [reply core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/LmdbSailStore.java:67]: Yep, looking at those benchmarks it seems to be an improvement across the board, so I can't immediately see a use case for _not_ wanting multi-threading enabled. Let's leave as-is for now - we can always decide to make it configurable later if we encounter a need for it.

**Review by jeenbroekstra — APPROVED** [review]

**Pull request merged**